### PR TITLE
Add current post type to withPluginContext to allow conditional rendering.

### PR DESCRIPTION
**Pull request by ryanwelcher**
## Description
Currently, there is no built-in way to determine the post type when using `PluginDocumentSettingPanel` to add a new panel to the document sidebar.

This pull request adds a new `currentPostType` prop that is passed to the children that will allow for conditional rendering.

## Types of changes
This introduces a new item being added via the existing withSelect HOC and checks if `children` are functions and passes the new property as a parameter. This _shouldn't_ cause any issues as if a function is passed now, a `Functions are not valid as a React child` is triggered.

## Checklist:
- [ ] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR. .

## Example Code
```
const { registerPlugin } = wp.plugins;
const { PluginDocumentSettingPanel } = wp.editPost;

const PluginDocumentSettingPanelDemo = () => (
	<>
		<PluginDocumentSettingPanel
			name={'my-custom-panel'}
			title={"Custom Panel Title"}
			className={"my-awesome-class"}
		>
			{ ( { currentPostType } ) => {
				if ( 'page' === currentPostType ) {
					<p>I am a Page!</p>
				} else {
					return (
						<p>{`Post Type: ${currentPostType}`}</p>
					)
				}
			}}

		</PluginDocumentSettingPanel>
	</>
);

registerPlugin( 'conditional-check', { render: PluginDocumentSettingPanelDemo } );

### CODE REVIEW

**Comment by ryanwelcher** [comment]
This PR closes #19392

**Comment by aduth** [comment]
I'm not necessarily opposed to this, but as far as answering the question, "Which information do we include?" the threshold is not very clear to me as far as what should be provided by default.

What makes it interesting, I think, is from #19392:

> This is possible using `withSelect` but having that data available by default would be a better solution.

... seems to imply that it is not very intuitive or simple to access this data, which itself could be seen to be a problem.

At least as far as conciseness, I don't see it being verbose using hooks:

```js
const currentPostType = useSelect( ( select ) => select( 'core/editor' ).getCurrentPostType() );
if ( 'page' === currentPostType ) {
    return null;
}

return <PluginDocumentSettingPanel />
```

This example also has me thinking: If, in your original example, you're returning `null` in the function child based on the post type, what are you expecting to be rendered for the setting panel? Is it expected to not be shown at all, or expected to be shown as empty? Based on the use-cases I could imagine for this sort of thing, it seems we might want the `PluginDocumentSettingPanel` to not be rendered at all, like in my example above.

File: packages/edit-post/src/components/sidebar/plugin-document-setting-panel/index.js [hunk 1]
```diff
@@ -10,14 +10,19 @@ import { compose } from '@wordpress/compose';
 import { withPluginContext } from '@wordpress/plugins';
 import { withDispatch, withSelect } from '@wordpress/data';
 
+/**
+ * External dependencies
+ */
```
Comment (aduth) [inline packages/edit-post/src/components/sidebar/plugin-document-setting-panel/index.js]: Per documented convention, can you place this grouping _before_ the "WordPress dependencies" ?

https://github.com/WordPress/gutenberg/blob/master/docs/contributors/coding-guidelines.md#imports

File: packages/edit-post/src/components/sidebar/plugin-document-setting-panel/index.js [hunk 2]
```diff
@@ -33,7 +38,7 @@ const PluginDocumentSettingFill = ( { isEnabled, panelName, opened, onToggle, cl
 						opened={ opened }
 						onToggle={ onToggle }
 					>
-						{ children }
+						{ ( isFunction( children ) ) ? ( children( { currentPostType } ) ) : ( children ) }
```
Comment (aduth) [inline packages/edit-post/src/components/sidebar/plugin-document-setting-panel/index.js]: The extra parentheses are unnecessary. I personally don't think they help add clarity, either.

```suggestion
						{ isFunction( children ) ? children( { currentPostType } ) : children }
```

**Comment by ryanwelcher** [comment]
@aduth 
> At least as far as conciseness, I don't see it being verbose using hooks:

100% agree here, it's just another step that needs to be done - albeit a small one.

> This example also has me thinking: If, in your original example, you're returning `null` in the function child based on the post type, what are you expecting to be rendered for the setting panel? Is it expected to not be shown at all, or expected to be shown as empty? Based on the use-cases I could imagine for this sort of thing, it seems we might want the `PluginDocumentSettingPanel` to not be rendered at all, like in my example above.

This is a very good point, and perhaps a bad usage example, as the example above won't stop rendering the Panel, to do that we'd need to control rendering outside the fill. It will however allow, different content to be displayed based on post type.

File: packages/edit-post/src/components/sidebar/plugin-document-setting-panel/index.js (hunk 1 above)
Comment (ryanwelcher) [reply packages/edit-post/src/components/sidebar/plugin-document-setting-panel/index.js]: Done.

File: packages/edit-post/src/components/sidebar/plugin-document-setting-panel/index.js (hunk 2 above)
Comment (ryanwelcher) [reply packages/edit-post/src/components/sidebar/plugin-document-setting-panel/index.js]: Updated

**Comment by gziolo** [comment]
> This is a very good point, and perhaps a bad usage example, as the example above won't stop rendering the Panel, to do that we'd need to control rendering outside the fill. It will however allow, different content to be displayed based on post type - I've updated the example to better reflect this.

You still can use React hooks at any level of nesting to perform the same checks. Is this something you feel is always used with this type of fills? We should follow the most common usage and expose props that are used in real-life implementations. If this prop is one of them then we definitely should include it by default, otherwise, we should add an example in the docs how to use hooks to make it easy to integrate.

**Comment by ryanwelcher** [comment]
> Is this something you feel is always used with this type of fills? We should follow the most common usage and expose props that are used in real-life implementations. If this prop is one of them then we definitely should include it by default, otherwise, we should add an example in the docs how to use hooks to make it easy to integrate.

Having access to the post type makes a lot of sense when trying to only add fills to a certain post type. @aduth makes a great point that this will only affect the content of the fill and not control whether or not it gets rendered. That being said, having access to the post type ( and/or perhaps other data ) makes sense here as well as the user may want to customize content based on post type.

Historically, when [adding meta boxes](https://developer.wordpress.org/reference/functions/add_meta_box/) to the Classic Editor, there was a parameter to control which screen(s) it was added. In addition, 
the callback that rendered the box also received the `$post` object. 

This is a different situation because we have access to the `select` API at any time but it would be a convenience to that this information available without any further work.

There is an argument for some kind of context that is available for all SlotFills that would provide information about the screen where it's being rendered.

**Comment by gziolo** [comment]
> Historically, when [adding meta boxes](https://developer.wordpress.org/reference/functions/add_meta_box/) to the Classic Editor, there was a parameter to control which screen(s) it was added. In addition, the callback that rendered the box also received the `$post` object.

That's a very good point to take into account 👍 



> > Is this something you feel is always used with this type of fills? We should follow the most common usage and expose props that are used in real-life implementations. If this prop is one of them then we definitely should include it by default, otherwise, we should add an example in the docs how to use hooks to make it easy to integrate.
> 
> Having access to the post type makes a lot of sense when trying to only add fills to a certain post type. @aduth makes a great point that this PR will only affect the content of the fill and not control whether or not it gets rendered. To your point, @gziolo, showing how to control rendering altogether may be better as an example in the docs.
> 
> That being said, having access to the post type ( and/or perhaps other data ) inside the fill makes sense here as well as the user may want to customize content based on post type.
> 
> Historically, when [adding meta boxes](https://developer.wordpress.org/reference/functions/add_meta_box/) to the Classic Editor, there was a parameter to control which screen(s) it was added. In addition, the callback that rendered the box also received the `$post` object.
> 
> This is a different situation because we have access to the `select` API at any time but it would be a convenience to that this information available without any further work.
> 
> There is an argument for some kind of context that is available for all SlotFills that would provide information about the screen where it's being rendered.

I would be in favor of exposing the same information as part of the plugin context so all fills could take advantage of it. This way it would be set only once and propagated to all fills.

**Comment by ryanwelcher** [comment]
> I would be in favor of exposing the same information as part of the plugin context so all fills could take advantage of it. This way it would be set only once and propagated to all fills.

This does seem to be a much cleaner and more scalable solution! I refactor this PR to get that happening.

As an aside here, while working on this PR, I discovered that any Fills that contain elements ( such as this one ) will need to be refactored to allow functions to be added to the implementation to check for props/fillprops - see https://github.com/WordPress/gutenberg/pull/19391/files#diff-dcb88e59de613bfe4c0eaa6dac1a13aeR42.

**Comment by ryanwelcher** [comment]
@gziolo I've updated this PR to add the post type to the plugin context instead of the Slot itself - any feedback would be greatly appreciated!

**Comment by aduth** [comment]
From my perspective, I'm having a hard time getting on board with this.

The purported benefit is that it makes values such as post type more accessible for a plugin author. As far as I can tell, it's the effective difference between:

```js
function MyPlugin( { currentPostType } ) {
    // ...
}
```

.. and:

```js
function MyPlugin() {
    const currentPostType = useSelect( ( select ) => select( 'core/editor' ).getCurrentPostType() );
    // ...
}
```

Granted, there's more code. However, consider:

- `currentPostType` is just one value. By contrast, `useSelect` here will grant access to any of the 333 state selectors currently available, plus derivations thereof.
- For anything other than the most trivial implementations, a plugin developer is inevitably going to need access to state. The above code is consistent with those expectations. `currentPostType` is custom to just the one or few fills which choose to provide it.
- There is already convention and automated processes in the [documentation of data selectors](https://developer.wordpress.org/block-editor/data/). For individual fill props, this process would be a manual, and thus could easily be overlooked (such as is the case in the current state of this pull request).

The main issues I have stem from maintainability and documentation, specifically: What is the metric against which we decide to include a default prop? And to which plugin fills are those applied? And how is this effectively communicated to plugin authors? Would they have any reason to _expect_ these props to exist? Would they be confused when a fill doesn't have a value they expect?

These will continue to be difficult questions to answer over time, and will most certainly lead to confusion, uncertainty, and indecision.

Instead, I'd much rather give plugin developers the tools to most easily access all or any of the specific requirements with a baseline and consistent toolkit (`useSelect` and store selectors). And if those aren't simple enough, then I think we should focus energies on making those more accessible.

> Historically, when [adding meta boxes](https://developer.wordpress.org/reference/functions/add_meta_box/) to the Classic Editor, there was a parameter to control which screen(s) it was added. In addition, the callback that rendered the box also received the `$post` object.

As mentioned previously, and if I understand correctly, the proposed changes would not help with this requirement, because at the point at which control is given to the fill, it's already too late to prevent anything from being rendered based on the post type (the wrapper will be rendered anyways).

The example was updated, but it's a different use-case than the one described here, and unlike this one is not obviously linked to a demonstrable need.

**Comment by chrisvanpatten** [comment]
Doesn't this also make the assumption that Plugin Context is only provided in a post-editing environment? I believe the current implementation is relatively agnostic.

I am _incredibly_ sympathetic to anything that improves extensibility of Gutenberg for third party developers, as someone who regularly finds himself banging his head against the wall wishing for a slotfill here or extra filters there or whatnot whenever my product team comes with some crazy/specific request.

However in this case, I agree with @aduth's conclusions — Gutenberg already provides intuitive APIs for accessing this data, and chances are good that you'll need to be hitting the data API anyway in your code (to set and retrieve post meta, make some kind of entity call, etc.). And when you add in the fact that this wouldn't even allow you to fully hide your fill (because the wrapper still renders) it just doesn't seem particularly valuable to an implementer.

**Comment by ryanwelcher** [comment]
Thanks for the feedback everyone! You make some excellent points and I think I’ll close this out.

**Pull request closed without merging**